storage/libradb/src/pruner/mod.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides `Pruner` which manages a thread pruning old data in the background and is
+//! meant to be triggered by other threads as they commit new data to the DB.
+
+use crate::{
+    schema::{
+        retired_state_record::RetiredStateRecordSchema, state_merkle_node::StateMerkleNodeSchema,
+    },
+    OP_COUNTER,
+};
+use failure::prelude::*;
+use logger::prelude::*;
+use schemadb::{ReadOptions, SchemaBatch, DB};
+use sparse_merkle::RetiredRecordType;
+use std::{
+    sync::{
+        mpsc::{channel, Receiver, Sender},
+        Arc, Mutex,
+    },
+    thread::JoinHandle,
+};
+use types::transaction::Version;
+
+#[cfg(test)]
+use std::thread::sleep;
+#[cfg(test)]
+use std::time::{Duration, Instant};
+
+const COUNTER_NAME: &str = "pruner.least_readable_state_version";
+
+/// The `Pruner` is meant to be part of a `LibraDB` instance and runs in the background to prune old
+/// data.
+///
+/// It creates a worker thread on construction and joins it on destruction. When destructed, it
+/// quits the worker thread eagerly without waiting for all pending work to be done.
+pub(crate) struct Pruner {
+    num_versions_to_keep: u64,
+    worker_thread: Option<JoinHandle<()>>,
+    command_sender: Mutex<Sender<Command>>,
+}
+
+impl Pruner {
+    /// creates a worker thread that waits on a channel for pruning commands.
+    pub fn new(db: Arc<DB>, num_versions_to_keep: u64) -> Self {
+        let (command_sender, command_receiver) = channel();
+
+        let worker_thread =
+            std::thread::spawn(move || Worker::new(db, command_receiver).work_loop());
+
+        Self {
+            num_versions_to_keep,
+            worker_thread: Some(worker_thread),
+            command_sender: Mutex::new(command_sender),
+        }
+    }
+
+    /// Sends pruning command to the worker thread when necessary.
+    pub fn wake(&self, latest_version: Version) {
+        if latest_version > self.num_versions_to_keep {
+            let least_readable_version = latest_version - self.num_versions_to_keep;
+            match self.command_sender.lock() {
+                Ok(sender) => sender
+                    .send(Command::Prune {
+                        least_readable_version,
+                    })
+                    .expect("Receiver should not destruct prematurely."),
+                Err(e) => crit!("Failed locking command_sender to pruner worker. {:?}", e),
```

Comment:
I think the convention is just to use `expect` and panic if the lock is poisoned.

Reply:
K. I was making the pruner thread as "low key" as possible (as in doesn't get in the way of other operations even if itself is failing.). But I guess this is a pretty big problem if it happens, that we don't want to swallow.